test/Identity.Test/IdentityServer/DeviceValidatorTests.cs

```diff
@@ -0,0 +1,223 @@
+﻿using Bit.Core.Context;
+using Bit.Core.Entities;
+using Bit.Core.Enums;
+using Bit.Core.Repositories;
+using Bit.Core.Services;
+using Bit.Core.Settings;
+using Bit.Identity.IdentityServer;
+using Bit.Test.Common.AutoFixture.Attributes;
+using Duende.IdentityServer.Validation;
+using NSubstitute;
+using Xunit;
+using AuthFixtures = Bit.Identity.Test.AutoFixture;
+
+namespace Bit.Identity.Test.IdentityServer;
+
+public class DeviceValidatorTests
+{
+    private readonly IDeviceService _deviceService;
+    private readonly IDeviceRepository _deviceRepository;
+    private readonly GlobalSettings _globalSettings;
+    private readonly IMailService _mailService;
+    private readonly ICurrentContext _currentContext;
+    private readonly DeviceValidator _sut;
+
+    public DeviceValidatorTests()
+    {
+        _deviceService = Substitute.For<IDeviceService>();
+        _deviceRepository = Substitute.For<IDeviceRepository>();
+        _globalSettings = new GlobalSettings();
+        _mailService = Substitute.For<IMailService>();
+        _currentContext = Substitute.For<ICurrentContext>();
+        _sut = new DeviceValidator(
+            _deviceService,
+            _deviceRepository,
+            _globalSettings,
+            _mailService,
+            _currentContext);
+    }
+
+    [Theory]
+    [BitAutoData]
+    public async void SaveDeviceAsync_DeviceNull_ShouldReturnNull(
+        [AuthFixtures.ValidatedTokenRequest] ValidatedTokenRequest request,
+        User user)
+    {
+        // Arrange
+        request.Raw["DeviceIdentifier"] = null;
+
+        // Act
+        var device = await _sut.SaveDeviceAsync(user, request);
+
+        // Assert
+        Assert.Null(device);
+        await _mailService.DidNotReceive().SendNewDeviceLoggedInEmail(
+            Arg.Any<string>(), Arg.Any<string>(), Arg.Any<DateTime>(), Arg.Any<string>());
+    }
+
+    [Theory]
+    [BitAutoData]
+    public async void SaveDeviceAsync_UserIsNull_ShouldReturnNull(
+        [AuthFixtures.ValidatedTokenRequest] ValidatedTokenRequest request)
+    {
+        // Arrange
+        request = AddValidDeviceToRequest(request);
+
+        // Act
+        var device = await _sut.SaveDeviceAsync(null, request);
+
+        // Assert
+        Assert.Null(device);
+        await _mailService.DidNotReceive().SendNewDeviceLoggedInEmail(
+            Arg.Any<string>(), Arg.Any<string>(), Arg.Any<DateTime>(), Arg.Any<string>());
+    }
+
+    [Theory]
+    [BitAutoData]
+    public async void SaveDeviceAsync_ExistingUser_NewDevice_ReturnsDevice_SendsEmail(
+        [AuthFixtures.ValidatedTokenRequest] ValidatedTokenRequest request,
+        User user)
+    {
+        // Arrange
+        request = AddValidDeviceToRequest(request);
+
+        user.CreationDate = DateTime.UtcNow - TimeSpan.FromMinutes(11);
+        _globalSettings.DisableEmailNewDevice = false;
```

Comment:
❓ : if we are going to keep the intro of this flag in this PR, then we should have a test where it is true and we assert the expected change in behavior.